R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
+    } else if (!(objective %in% .MULTI_TARGET_OBJECTIVES())) {
+      stop(
+        "Got multi-column 'y' - supported objectives for this data are: ",
+        paste(.MULTI_TARGET_OBJECTIVES(), collapse = ", "),
+        ". Was passed: ",
+        objective
+      )
+    }
+
+    y_names <- names(y)
+    y <- lapply(y, function(x) {
+      if (!inherits(x, c("numeric", "integer"))) {
+        stop(
+          "Multi-target 'y' only supports 'numeric' and 'integer' types. Got: ",
+          paste(class(x), collapse = ", ")
+        )
+      }
+      return(as.numeric(x))
+    })
+    y <- as.data.frame(y) |> as.matrix()
+
+    if (!is.null(base_margin)) {
+      base_margin <- process.base.margin(base_margin, length(y), ncol(y))
+    }
+
+    out <- list(
+      params = list(
+        objective = objective
+      ),
+      dmatrix_args = list(
+        label = y,
+        base_margin = base_margin
+      ),
+      metadata = list(
+        y_names = y_names,
+        n_targets = ncol(y)
+      )
+    )
+
+  } else if (is.matrix(y)) {
+    if (ncol(y) == 1L) {
+      return(process.y.margin.and.objective(as.vector(y), base_margin, objective, params))
+    }
+
+    if (!is.null(objective) && !(objective %in% .MULTI_TARGET_OBJECTIVES())) {
+      stop(
+        "Got multi-column 'y' - supported objectives for this data are: ",
+        paste(.MULTI_TARGET_OBJECTIVES(), collapse = ", "),
+        ". Was passed: ",
+        objective
+      )
+    }
+    if (is.null(objective)) {
+      objective <- "reg:squarederror"
+    }
+
+    y_names <- colnames(y)
+    if (storage.mode(y) != "double") {
+      storage.mode(y) <- "double"
+    }
+
+    if (!is.null(base_margin)) {
+      base_margin <- process.base.margin(base_margin, nrow(y), ncol(y))
+    }
+
+    out <- list(
+      params = list(
+        objective = objective
+      ),
+      dmatrix_args = list(
+        label = y,
+        base_margin = base_margin
+      ),
+      metadata = list(
+        n_targets = ncol(y)
+      )
+    )
+
+    if (NROW(y_names) == ncol(y)) {
+      out$metadata$y_names <- y_names
+    }
+
+  } else {
+    stop("Passed 'y' object with unsupported class: ", paste(class(y), collapse = ", "))
+  }
+
+  return(out)
+}
+
+process.row.weights <- function(w, lst_args) {
+  if (!is.null(w)) {
+    if ("label" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- NROW(lst_args$dmatrix_args$label)
+    } else if ("label_lower_bound" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- length(lst_args$dmatrix_args$label_lower_bound)
+    } else {
+      stop("Internal error.")
+    }
+    if (!is.numeric(w)) {
+      w <- as.numeric(w)
+    }
+    if (length(w) != nrow_y) {
+      stop(
+        "'weights' must be a 1-d vector with the same length as 'y' (",
+        length(w), " vs. ", nrow_y, ")."
+      )
+    }
+    lst_args$dmatrix_args$weight <- w
+  }
+  return(lst_args)
+}
+
+check.nthreads <- function(nthreads) {
+  if (is.null(nthreads)) {
```

Comment:
Is this going to make single thread the default for R xgboost?

Reply:
No, the default is to use all threads, as obtained when calling `parallel::detectCores()`.